Language library refactor

graceyim opened the pull request:
Updated existing toopher-java library using blueprint v3

Commit:
Remove pairUri()

Commit:
Refactor authUri() and update tests

Commit:
Update javadocs comment for allowInlinePairing

Commit:
Create getUserManagementUrl() with test

Commit:
Rename uri to url

Commit:
Refactor getAuthenticationUrl() so that it treats extras as a black box

Commit:
Rename PairingStatus to Pairing

Commit:
Update javadocs for extras

Commit:
Rename validate to validatePostback

Commit:
Create User class and update Pairing and tests

Commit:
Add comment for old test

Commit:
Rename getPairingStatus to getPairing

Commit:
Use JSONObject for HttpClientMock

Commit:
Replace ToopherAPI.getPairing() with ToopherAPI.advanced.pairings.getById()

Commit:
Correct javadocs for setToopherEnabledForUser()

Commit:
Rename AuthenticationStatus to AuthenticationRequest

Commit:
Create UserTerminal class

Commit:
Create ToopherAPI.advanced.authenticationRequests.getById()

Commit:
Create ToopherAPI.advanced.users.getById()

Commit:
Add variables to setUp() and DEFAULT_BASE_URL for tests

Commit:
Cleanup ToopherAPI, TestToopherAPI and User

Commit:
Create ToopherAPI.advanced.users.getByName()

Commit:
Create ToopherAPI.advanced.users.create()

Commit:
Fix endpoint in Users.getByName()

Commit:
Create ToopherAPI.advanced.userTerminals.getById()

Commit:
Create Action object

Commit:
Add Action and User to AuthenticationRequest object

Commit:
Remove raw from UserTerminal

Commit:
Create ApiRawRequester class for get() and post()

Commit:
Refactor pair() to handle regular, QR and SMS pairings

Commit:
Add terminalNameExtra to tests

Commit:
Remove old getAuthenticationStatus()

Commit:
Refactor authenticate() to handle authentication with username or pairing id

Commit:
Add AuthenticationRequest.authenticate_with_otp()

Commit:
Add update() to Action

Commit:
Add update() and remove raw from User

Commit:
Add update() to UserTerminal

Commit:
Update AuthenticationRequest with update methods

Commit:
Add AuthenticationRequest.refreshFromServer()

Commit:
Add toString() for UserTerminal

Commit:
Update User when updating UserTerminal

Commit:
Add UserTerminal.refreshFromServer()

Commit:
Remove raw from Pairing

Commit:
Add Pairing.refresh_from_server() and update()

Commit:
Make update() private in AuthenticationRequest and Pairing

Commit:
Change DEFAULT_TTL to 300

Commit:
Add Pairing.getResetLink()

Commit:
Add Pairing object to test setUp()

Commit:
Add Pairing.emailResetLink()

Commit:
Add extras to params in Pairing.getResetLink()

Commit:
Use ApiRawRequester.post(endpoint, params) when extras are null

Commit:
Add ApiRawRequester.get(endpoint, params) for when extras are null

Commit:
Update HttpClientMock and add ResponseMock to mock 2 responses to test Users.getByName

Commit:
Rename endpoint for UserTerminal.getById to be consistent with other endpoints

Commit:
Rename DEFAULT_BASE_URL endpoint in tests to be consistent with ToopherAPI url format

Commit:
Add reason_code to AuthenticationRequest

Commit:
Use camelCase, not snake_case

Commit:
Add api to User(s), UserTerminal(s), AuthenticationRequest(s) and Pairing(s)

Commit:
Rename authenticateWithOtp to grantWithOtp

Commit:
Fix testAdvancedUsersCreateWithExtras()

Commit:
Rename User.enabled to User.toopherAuthenticationEnabled

Commit:
Update setter for User.toopherAuthenticationEnabled

Commit:
Use camelCase, not snake_case

Commit:
Add User.refreshFromServer()

Commit:
Add User.enableToopherAuthentication()

Commit:
Add User.disableToopherAuthentication()

Commit:
Add User.reset()

Commit:
Rename raw to rawResponse in ApiResponseObject

Commit:
Clean up javadocs

Commit:
Add Pairing.getQrCodeImage()

Commit:
Use String.format for strings

Commit:
Use DEFAULT_BASE_URL in ToopherAPI tests

Commit:
Clean up tests for Pairing

Commit:
Clean up tests for AuthenticationRequest

Commit:
Clean up tests for AuthenticationRequest

Commit:
Update HttpClientMock.getLastCalledEndpoint() to return endpoint without DEFAULT_BASE_URL

Commit:
Update HttpClientMock.getLastCalledEndpoint() to return endpoint without DEFAULT_BASE_URL

Commit:
Remove methods from ToopherAPI that have been moved to AuthenticationRequest and User

Commit:
Add UserTerminals.create to replace assignUserFriendlyNameToTerminal

Commit:
Clean up ApiRawRequester.get() and post()

Commit:
Update HttpClientMock.getExpectedResponse to return expectedResponseBody if expectedUriResponses is null

Commit:
Clean up and add tests for ToopherAPI

Commit:
Merge branch 'lang-library-refactor' of github.com:toopher/toopher-java into lang-library-refactor

Conflicts:
	src/test/java/com/toopher/TestAuthenticationRequest.java

Commit:
Clean up tests for ToopherIframe

Commit:
Remove leading / from User.reset endpoint

Commit:
Initialize json responses in @BeforeClass

Commit:
Clean up and add tests for User

Commit:
Clean up and add tests for User

Commit:
Clean up and add tests for UserTerminal

Commit:
Use @BeforeClass instead of @Before for TestAuthenticationRequest

Commit:
Add javadocs and clean up UserTerminal

Commit:
Add ApiResponseObject.updateRawResponse for updates

Commit:
Add javadocs and clean up User

Commit:
Add javadocs and clean up Pairing

Commit:
Update javadocs and clean up Pairing

Commit:
Update javadocs and clean up ToopherAPI

Commit:
Update javadocs and clean up AuthenticationRequest

Commit:
Update ToopherAPIDemo

Commit:
Update ApiResponseObject.updateRawResponse to throw JSONException

Commit:
Update Action.update to throw JSONException

Commit:
Update AuthenticationRequest to throw JSONException

Commit:
Update Pairing to throw JSONException

Commit:
Update ToopherAPI to throw JSONException

Commit:
Update ToopherAPIDemo to catch JSONException

Commit:
Update User to throw JSONException

Commit:
Update UserTerminal to throw JSONException

Commit:
Update README with new pair() and authenticate()

Commit:
Update Iframe README

Commit:
Rename validate to validatePostback

Commit:
Rename ToopherAPI to ToopherApi

Commit:
Use 'this' in AdvancedApiUsageFactory

Commit:
Use 'this' in AdvancedApiUsageFactory

smholloway commented:
Can you fix the build error, please?
```
toopher-java/src/main/java/com/toopher/ToopherAPI.java:[37,7] class ToopherApi is public, should be declared in a file named ToopherApi.java
```

smholloway commented on src/main/java/com/toopher/AuthenticationRequest.java:
```diff
@@ -0,0 +1,132 @@
+     */
+    public ToopherApi api;
+
+	/**
```
Out of alignment.

smholloway commented on src/main/java/com/toopher/AuthenticationRequest.java:
```diff
@@ -0,0 +1,132 @@
+
+/**
+ * Provides information about the status of an authentication request
+ * 
```
Blank line.

smholloway commented on src/main/java/com/toopher/AuthenticationRequest.java:
```diff
@@ -0,0 +1,132 @@
+    /**
+     * The UserTerminal associated with the authentication request
+     */
+    public UserTerminal terminal;
```
Could you look into using links to the objects in the JavaDoc? It's something like `{@link com.toopher.UserTerminal}`

graceyim replied:
Done!

smholloway commented on src/main/java/com/toopher/AuthenticationRequest.java:
```diff
@@ -0,0 +1,132 @@
+    public User user;
+
+    public AuthenticationRequest(JSONObject json, ToopherApi toopherApi) throws JSONException{
+		super(json);
```
Out of alignment.

smholloway commented on src/main/java/com/toopher/Action.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public void update(JSONObject jsonResponse) throws JSONException {
+        this.name = jsonResponse.getString("name");
```
Are we missing a `rawResponse` object in this class? In `AuthenticationRequest#update` I see `this.updateRawResponse(jsonResponse);` but a similar statement is not present here.

graceyim replied:
I didn't add `updateRawResponse()` since `Action.raw` wasn't in the blueprint, but I can add it so it's consistent across all objects that extend ApiResponseObject

smholloway commented on src/main/java/com/toopher/ToopherAPI.java:
```diff
@@ -153,362 +151,145 @@ public ToopherAPI(String consumerKey, String consumerSecret, URI uri, HttpClient
-     *
-     * @param pairingPhrase
-     *            The pairing phrase supplied by the user
+     * Create a QR pairing
```
This is a big vague. I think we'll want to add some additional information, like "This return a PNG that should be scanned by the user to complete the pairing process."

Commit:
Merge branch 'lang-library-refactor' of github.com:toopher/toopher-java into lang-library-refactor

Commit:
Fix spacing in AuthenticationRequest

Commit:
Rename ToopherAPI/TestToopherAPI to ToopherApi/TestToopherApi

Commit:
Update rawResponse for Action

Commit:
Use old style of testing authenticationUrl by assertEquals oauth_signature

smholloway commented on src/test/java/com/toopher/HttpClientMock.java:
```diff
@@ -18,16 +18,23 @@
     private int expectedResponseStatus;
     private String expectedResponseBody;
+    private Map<URI, ResponseMock> expectedUriResponses;
+    private java.net.URI lastURI;
```
I think you can drop the `java.net.` piece and just declare type `URI`.

graceyim replied:
Yep! Change made.

Commit:
Reformat all com.toopher files

Commit:
Use URI instead of java.net.URI

Commit:
Add links to objects in javadocs

smholloway commented on src/test/java/com/toopher/HttpClientMock.java:
```diff
@@ -65,22 +100,22 @@ public String getLastCalledData(String key) {
-            result = null;
-            e.printStackTrace();
-        }
+        result = responseHandler.handleResponse(resp);
```
I think you can combine the declaration and initalization: `T result = responseHandler.handleResponse(resp);`

graceyim replied:
Yep! Just fixed this.

smholloway commented on src/test/java/com/toopher/TestAuthenticationRequest.java:
```diff
@@ -0,0 +1,122 @@
+        newJsonResponse.remove("pending");
+        newJsonResponse.remove("granted");
+        newJsonResponse.put("pending", false);
+        newJsonResponse.put("granted", true);
```
Can we not set these values directly? Seems easier to do an update than a delete and create combo.

graceyim replied:
It looks like put() will replace, so no need for remove().

smholloway commented on src/test/java/com/toopher/TestToopherAPI.java:
```diff
@@ -1,89 +1,363 @@
-        PairingStatus pairing = toopherApi.getPairingStatus("1");
+        ToopherApi toopherApi = new ToopherApi("key", "secret",
+                createURI(DEFAULT_BASE_URL), httpClient);
+        Pairing pairing = toopherApi.pair(userName, "123456");
```
Can you use a string that looks more like a real phone number? E.g., `"5245678"` or `"5551234"`

smholloway commented on src/test/java/com/toopher/TestToopherAPI.java:
```diff
@@ -1,89 +1,363 @@
+
+        HttpClientMock httpClient = new HttpClientMock(200, userJsonResponse.toString());
+        ToopherApi toopherApi = new ToopherApi("key", "secret",
+                createURI(DEFAULT_BASE_URL), httpClient);
```
How do you feel about putting the `toopherApi` creation into the common test setup? (It seems to be repeated in every test.)

Commit:
Add User.toopher_authentication_enabled and remove disable_toopher_auth

Commit:
Rename terminal_name_extra to requester_specified_terminal_id

Commit:
Rename UserTerminal.name_extra to requester_specified_id

dshafer commented on README-Iframe.md:
```diff
@@ -104,78 +101,8 @@ In this example, `data` is a `Map<String, String[]>` of the form data POSTed to
-
-    try {
-        Map<String, String> validatedData = iframeApi.validate(data, requestToken);
-        if (validatedData.containsKey("error_code")) {
```
Are we eliminating this from our iframe readme entirely? Seems like if we include any documentation about the two-stage iframe workflow, this part of the docs will be necessary.

graceekyim replied:
I meant to move this up to Step 2: Validate the result. I'll update this doc and add this back in. Thanks for catching this.

graceyim replied:
This block was moved to lines 72-102. If you look [here](https://github.com/toopher/toopher-java/blob/lang-library-refactor/README-Iframe.md), it's easier to read.

dshafer commented on src/main/java/com/toopher/ToopherAPIDemo.java:
```diff
@@ -1,149 +1,165 @@
-				if (pairingStatus.enabled) {
-					System.out.println("Pairing complete");
+
+        String pairingId;
```
If we're pushing developers toward authenticating by username (née Zero Requester Storage), shouldn't we update the demo to use that as well?

dshafer commented:
Should we also implement `ToopherPairingDeactivatedError`?  It would only be relevant when authenticating by pairingId, but still, it's an error that we do return...

Commit:
Update javadocs for ToopherIframe.getUserManagementUrl

Commit:
Update ToopherAPIDemo

Commit:
Combine declaration and initialization of result

Commit:
No need to remove before put on JSONObject

Commit:
Add ToopherPairingDeactivatedError to parseRequestError

Commit:
Update testCreateSmsPairing to use number formatted as phone number

Commit:
Use getToopherApi() to create new ToopherApi

Commit:
Return new User from Users.getByName instead of getById

Commit:
Remove expectedUriResponses since Users.getByName only needs one response

Commit:
Fix ant build errors

Commit:
Check if resetEmail is null before adding to params

Commit:
Update README

Commit:
Cleanup DEMO text

Commit:
Replace deprecated DefaultHttpClient

Commit:
Add demo command for mvn

Commit:
Remove unnecessary printlns

Commit:
Refactor ToopherApi.authenticate to accept terminalName and requesterSpecifiedId as args

Commit:
Test ToopherIframe.dateOverride and nonceOverride

Commit:
Add CONTRIBUTING.md for dependency and test info

Commit:
Add ToopherIframe test for validatePostback without TTL

Commit:
Add method overrides for getAuthenticationUrl and add tests

Commit:
Add test for ToopherIframe creation and getDate

Commit:
Add default resetEmail for getUserManagementUrl and test

Commit:
Add javadocs for getAuthenticationUrl

Commit:
Add tests for parseRequestError

Commit:
Test oauth_signature when extras are passed to getAuthenticationUrl

Commit:
Add tests for Users.getByName

Commit:
Remove unused method overrides

Commit:
Add test for AuthenticationRequest.toString

Commit:
Add test for User.toString

Commit:
Add test for UserTerminal.toString

Commit:
Add test for Pairing.toString

Commit:
Update pom to exclude ToopherAPIDemo from jacoco reports

Commit:
Add tests for ToopherApi create

Commit:
Expand tests for ToopherApi.authenticate

Commit:
Cleanup ToopherApi.authenticate

Commit:
Add tests for parseRequestError

Commit:
Add test for empty QR response body

Commit:
Make sure tests fail when expected error is not thrown

Commit:
Test ResponseHandler when response body is empty

Commit:
Add another test for getAuthenticationUrl

Commit:
Assert error codes are correct for ToopherClientErrors

Commit:
Add method override for ToopherIframe.getUserManagementUrl

Commit:
Add override for ToopherIframe.validatePostback and test

Commit:
Rename validatePostback to processPostback

Commit:
Add getExtrasForUrl helper method for tests

Commit:
Add getPostbackData helper method for tests

Commit:
Move ToopherIframe.setDateOverride to test setUp

Commit:
Update tests to throw SignatureValidationError instead of try/catch block

Commit:
Update ToopherApi tests

Commit:
Refactor getAuthenticationUrl and tests

Commit:
Use processPostback to verify authenticity of ToopherIframe postback data

Commit:
Use Iframe instead of iframe

Commit:
Add isAuthenticationGranted and tests

Commit:
Add ttl and expires to params in getOAuthUrl

Commit:
Set defaults for resetEmail, sessionToken and requesterMetadata as empty strings

Commit:
Add resetEmail to extras for getUserManagementUrl

Commit:
Use a map for Toopher iframe postback data

Commit:
Increase test coverage for ToopherIframe

Commit:
Merge branch 'develop' of github.com:toopher/toopher-java into lang-library-refactor

Conflicts:
	CONTRIBUTING.md

Commit:
Update js for ToopherIframe postback data

Commit:
Update demo to match toopher-python

Commit:
Include baseUri when creating new instance of ToopherApi

Commit:
Add MIT license

Commit:
Update ToopherIframe tests to throw URISyntaxException for processPostback

Commit:
Cleanup javadocs

Commit:
Refactor ToopherApi.authenticate

Commit:
Reorder methods in ToopherIframe

Commit:
Remove unused exception

Commit:
Remove unused imports

Commit:
Add license to README

Commit:
Add processPostback test for keys with empty values

Commit:
Simplify README and add iframe info

dshafer committed:
Bump pom.xml version to 2.0.0-PREVIEW

dshafer committed:
Add jar-with-dependencies goal

To build a single jarfile with all dependencies, run:
`mvn assembly:assembly -DdescriptorId=jar-with-dependencies`
